register derived conditions before reading the savefile; and set the …

leklachu opened the pull request:
Edit: the finished version does not adjust when the "reputation with" is applied, meaning reputation in that overrides obsolete reputation in conditions. It also does not touch new games, only loaded games.
- If anyone comes across this because their save is corrupted from a mission reputation change that didn't apply, like https://github.com/endless-sky/endless-sky/issues/7217 , you can either load an earlier save than that mission, then this fix should work for you, or you can edit your savefile, then use this fix.

---

Possible solution to https://github.com/endless-sky/endless-sky/issues/7217

This change
- Registers the derived conditions before reading the savefile (or the `start` for a new player)
- When reading the saved "reputation with", stores directly to the GameData government reputation, instead of queuing to save later.

The result
- Any saved primary `condition`s that belong in a derived condition, will load into their correct derived condition setter, then the obsolete primary condition will be forgotten and lost from the savefile
- For reputation, the old conditions (if still there) will overwrite the `"reputation with"` block
- Government reputations will now be all rounded to integers, because the conditions were
- Therefore **this should fix** people's games whether before they hit a reputation-changing mission, or if they've already had one.
  - I haven't looked at what happens if a mission *adjusts* instead of *setting* reputation
- It should also stop future obsolete primary conditions from perma-bugging the game when new derived conditions are implemented
  - but care should be given to where the 'correct' data comes from at the point of the change

Non-results
- Any other derived condition, which is temporarily loaded then applied through `ApplyChanges()` (as reputation was), will overwrite the obsolete conditions with whatever is loaded, effectively ignoring and removing the obsolete conditions instead of applying then removing.
- The warning suggested by @petervdmeer is not implemented. Perhaps with this change it's not needed

Assumptions
- Since mission actions currently affect primary conditions if present, any primary condition reputations will be more correct than their `"reputation with"` counterparts
  - this will wipe any reputation changes other than by mission, since the derived-condition-PR came into effect. Surely mission actions are more important
- `"reputation with"` in the save is always before `"conditions"`. Otherwise, `"reputation with"` will take priority if both provide reputations.
- fractional reputations are not that important, and it's okay to truncate them when making this fix.
- Derived conditions are only ever registered the once: on new/load player
- The savefile is only read once

Testing
- I've tested this on my out-of-sync save, which had Syndicate reputation stuck on ~400 but condition `"reputation: Syndicate" -1000` already set from the mission
- I haven't tested on a new game

Style
- Seems like my Vim, or something, did a bunch of style changes (including the copyright header?!)
- If we do approve this solution, perhaps it's easiest if I just re-do it (it's such a small edit) and make sure not to change other formatting, so as not to clutter edits in git's record

Correctness
- I've made this PR so you can see and try. It needs really carefully checking before being part of the main game
- Both changes I've made, the old way (or the comments) implied it was deliberately done in that order, which I've 'concluded logically' isn't necessary
- I can't promise I haven't overlooked something.
- (I will however continue to look through with this, if we don't simply reject this proposed solution)

Haste
- Any solution to this shouldn't be implemented without careful checking, lest we make this worse!
- But this is a serious bug, stopping people from progressing properly in any campaign that sets reputation, so I think it should be fixed ASAP

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -7,10 +7,7 @@ Foundation, either version 3 of the License, or (at your option) any later versi
-
-You should have received a copy of the GNU General Public License along with
-this program. If not, see <https://www.gnu.org/licenses/>.
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Looks like some coding style fixes (from https://github.com/endless-sky/endless-sky/pull/7216 and another PR) got lost in this PR (probably due to a failed rebase, or by overwriting the file with an older version).
Please make sure that the work from those other PRs doesn't get lost.

leklachu replied:
Okay, have manually fixed those lines to match with Master. Is that okay? The diff-with-master should now only be my changes.

petervdmeer replied:
Yes, that is fine. Thanks!

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -107,8 +104,10 @@ void PlayerInfo::New(const StartConditions &start)
+	// Register the derived conditions before applying start conditions, so that any that belong in a derived provider go into that correctly.
 	RegisterDerivedConditions();
+	start.GetConditions().Apply(conditions);
+	// RegisterDerivedConditions() used to be here
```
This fix looks good. Please just remove the out-commented code to keep the code-base clean. (Also for out-commented code later.)

leklachu replied:
Removed. My comments are still a bit verbose: wasn't sure where's the best balance.

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -177,8 +179,10 @@ void PlayerInfo::Load(const string &path)
+					// reputationChanges.emplace_back(
+					// 	GameData::Governments().Get(grand.Token(0)), grand.Value(1));
+					GameData::Governments().Get(grand.Token(0))->
+						SetReputation(grand.Value(1));
```
The original author of the code here had a reason to apply the reputation Changes later (in ApplyChanges). I'm not sure why, but I guess there was a good reason to delay applying those changes. Do you have an idea on why this applying of the changes was delayed?

And is there a reason not to do this? What if we just keep applying those reputationChanges later (because then they also overwrite any reputationChanges that were set from the conditions)?

leklachu replied:
That still worries me. I can't see the logic of it, thus I may have missed something important. Unless:
- it was done because the mechanism for deferred-application was made and tested, so doing it that way felt neat and tidy
- or it mattered before the change to derived conditions, perhaps making sure condition-reputation was freshly updated from game-reputation, if the duplicator hadn't run

Could some of the conditions that get deferred-applied be operations to modify conditions? Seems strange to have any of those at load-time though. At least the reputation ones are just listed as absolute.

Reason not to do this? Specifically so old primary-condition-reputations take precedence, because they'll only exist if it's a savegame from the old system, and if any mission has set them they'll be more 'correct' than the gamedata-stored ones.
- otherwise if a mission changes your reputation (such as ceasing hostilities with Republic) before applying this fix, that change will be lost and you'll be forever hated by the republic

petervdmeer replied:
> * or it mattered before the change to derived conditions, perhaps making sure condition-reputation was freshly updated from game-reputation, if the duplicator hadn't run

Yes, or maybe this was done because:
- other government data needed to be loaded before applying the reputations.
- mission loading somehow affects reputations, thus the actual reputations need to be set after loading the missions
- the default (standard) reputations from the data-files are loaded late in the process, causing the reputations from the savegame to be overwritten by the standard reputations from the data files.

ES uses "delayed loading" at more locations, and usually there was a very good reason for it due to interaction with other data or missing data. The reason here might no longer be valid, but we need to be really sure that such reason is no longer valid before I would be okay with this change. Just assuming that it will be no problem is not okay here, especially since assumptions (by me) on how loading works is what got us into this bug-fixing in the first place.

petervdmeer commented on source/PlayerInfo.cpp:
```diff
@@ -1203,8 +1207,7 @@ void PlayerInfo::Land(UI *ui)
 	{
-		string message = "These active missions or jobs were deactivated due to a missing definition"
-			" - perhaps you recently removed a plugin?\n";
+		string message = "These active missions or jobs were deactivated due to a missing definition - perhaps you recently removed a plugin?\n";
```
Some style fixes also appear to be lost here and on the lines below. Please ensure that the fixes from other PRs are not lost.

petervdmeer replied:
Fixed in follow-up commit.

petervdmeer reviewed (changes requested):
Thanks for working on this fix. Good to see this being picked up.

petervdmeer commented:
> * The warning suggested by @petervdmeer is not implemented. Perhaps with this change it's not needed

Keeping the warning out of _this_ PR is fine to keep this PR focused on the minimal fix to get the issue out of the way.

> * Seems like my Vim, or something, did a bunch of style changes (including the copyright header?!)

I don't think it was an editor like Vim, I think somehow in the process the PlayerInfo.cpp file got overwritten by an older version.

> * If we do approve this solution, perhaps it's easiest if I just re-do it (it's such a small edit) and make sure not to change other formatting, so as not to clutter edits in git's record

We perform Squash-merging for ES, so adding another commit in this PR is not going to affect the main/master branch. Just fixing it here in this PR seems most appropriate (to avoid having 2 commits on main/master where 1 would have been sufficient).

> Since mission actions currently affect primary conditions if present, any primary condition reputations will be more correct than their `"reputation with"` counterparts
> 
>     * this will wipe any reputation changes other than by mission, since the derived-condition-PR came into effect. Surely mission actions are more important

This argument only applies to people that are currently running the game on the HEAD of the master branch. Backwards compatibility is valuable, but I feel the more correct solution (of just applying the `"reputation with"` in the `ApplyChanges()` function) should take presence here, because the majority of the playerbase will play the stable or unstable released version.

(And yes, I feel very sorry for the people whose savegames I broke, but I still feel we should prioritize the correctness for the release versions over the backwards compatibility of HEAD-of-branch versions.

> Haste
> 
>     * Any solution to this shouldn't be implemented without careful checking, lest we make this worse!
>     * But this is a serious bug, stopping people from progressing properly in any campaign that sets reputation, so I think it should be fixed ASAP

Agreed on both. Let's fix this the proper way (for the release versions) and merge this soon to minimize the risk of others getting hit by the missing reputations issue.

leklachu commented:
I've updated the PR to restore the lost-style-changes. I have one more concern before it's applied: that's the logic in the new game. I can't figure out if applying conditions through a ConditionSet, as is done with `start` (L115, `start.GetConditions().Apply(conditions);`, respects derived conditions properly. I presume it does because it surely comes up elsewhere. If you know, that'd save me digging till I find out for sure.

(I'd also like to dig through ApplyChanges in Load(), but I'm satisfied that what I've changed is working correctly.)

leklachu commented:
Tested loading my out-of-sync save, and diff old-save/autofixed-save

The changes are:
- reputations are all set to what the conditions were at time of fix
  - losing sub-integer precision, and the minor changes from fighting pirates/etc
  - but restoring the major changes made by missions
- Lost `flagship model`, `flagship planet` and `flagship system` from the primary conditions: I presume those are derived now
- Lost all the reputation conditions, as expected
- Lost three `ships: ` primary conditions. Also presumably derived now.

petervdmeer commented:
> I've updated the PR to restore the lost-style-changes. I have one more concern before it's applied: that's the logic in the new game. I can't figure out if applying conditions through a ConditionSet, as is done with `start` (L115, `start.GetConditions().Apply(conditions);`, respects derived conditions properly. I presume it does because it surely comes up elsewhere. If you know, that'd save me digging till I find out for sure.

This sounds like a very good addition to the [unit-tests](https://github.com/endless-sky/endless-sky/tree/master/tests/unit/src); derived conditions in combination with ConditionSets. We don't have those tests at the moment. But feel free to use the fastest method of verification for this PR (even if that is manual), the unit-tests can be added after the fix in this PR.

> (I'd also like to dig through ApplyChanges in Load(), but I'm satisfied that what I've changed is working correctly.)

Good to hear.

petervdmeer commented:
>     * reputations are all set to what the conditions were at time of fix
>       * losing sub-integer precision, and the minor changes from fighting pirates/etc
>       * but restoring the major changes made by missions

This is nice to have, but I consider the current implementation risky (see my other review comments). I prefer to use the more safe route and just keep setting the reputations from the `"reputation with"` sections. On a related note: I like the fix in this PR to be merged very soon, to stop new persons from getting affected by this (if they are running their games from the latest master branch).

>     * Lost `flagship model`, `flagship planet` and `flagship system` from the primary conditions: I presume those are derived now

They are indeed derived now.
 
>     * Lost all the reputation conditions, as expected

Good.

>     * Lost three `ships: ` primary conditions. Also presumably derived now.

They are indeed derived now.

leklachu commented:
> the more safe route

...reluctantly agreed. Reluctant because anyone in my position will have a game-breaking bug. But hopefully those who've hit this situation within these few days will be few enough and can file a bug report (or see mine) and find the answer to fix their saves.

But the more I think about it, I realise my assumption that (soon-to-be-)obsolete primary conditions are usually correct over other storage places is not sound.

With this fix, your warnings from the other PR won't trigger; the obsolete primaries will be silently fitted in to the deriveds, then any deferred entries will overwrite those. Maybe a future PR could swap the order back, and add logic that, on adding a prefix where there's already primaries in its range, will warn, backup the save, add-in the in-range conditions into the prefix, and remove the primary conditions.

> derived conditions in combination with ConditionSets

I'm not really sure how to test this. I'll try and look through the code today, to see if it looks right. Or, I can just swap the order back for a new game: then at least it should give warnings (from your PR) if ever a start condition gets nuked by a derived prefix.

I'll also check `ApplyChanges()` in `Load()`, because that's a similar swapped order that I haven't checked right through (but at least my save-diff says it looks right for now!...) It used to `ApplyChanges` before registering derived conditions.

leklachu commented:
Possible other bug:

`PlayerInfo::GetSubstitutions()` ~L1827, gets substitutions from the conditions, including Flagship, which is now a derived condition. It looked like it's checking primary conditions...

I'd better not check this now, or I won't get the current bugfix checked before I might be unavailable a few days. But thought I'd jot it here as related so someone can check later.

leklachu commented:
Okay, I looked through.

- During the read-save-file section, there is no direct use of `conditions` besides the `conditions` block, and no indirect use I could see. So the move of `RegisterDerivedConditions` should cause no trouble there
- in `ApplyChanges`, almost nothing touches the conditions (including the reputations: because those go straight into GameData). That should be fine
- One thing does: `tribute: `. If that ever becomes a derived condition, it needs to be fixed here or it will probably overwrite the derived version. For now, it's fine.

Reputation loading priority
- have reverted the behaviour so `reputation with` defers application and overrides `conditions .. reputation: _` (see discussion above)

New saves
- `start...ApplyConditions()` I'm pretty sure is just setting values directly into `ConditionsStore::storage` map as if derived conditions and safety functions don't exist. If an alternate start sets any derived conditions, this will probably be a bug.
- I reverted the order of `RegisterDerivedConditions` in new game, because changing the order does not solve the bug and may make it worse. At least this way round, https://github.com/endless-sky/endless-sky/pull/7226 should throw warnings

With that done, this is a tiny change to code, with big implications and much examination. I believe this is ready to commit. I hope so, because I might not be available over the weekend!
